packages/commerce-server/src/format-prices-for-product.ts

```diff
@@ -298,6 +296,35 @@ export async function formatPricesForProduct(
         upgradedProduct,
       }),
     }
+  } else if (hasPurchaseWithPPP && upgradeFromPurchase) {
+    const {getMerchantCoupons} = getSdk({ctx})
+    const merchantCoupons =
+      (await getMerchantCoupons({
+        where: {type: 'ppp', percentageDiscount: pppDiscountPercent},
+      })) || []
+
+    await getMerchantCoupon({where: {id: merchantCouponId}})
+
+    const {identifier, merchantAccountId, ...merchantCouponWithoutIdentifier} =
+      merchantCoupons[0]
+
+    return {
+      ...defaultPriceProduct,
+      calculatedPrice: getCalculatedPriced({
+        unitPrice: defaultPriceProduct.unitPrice,
+        percentOfDiscount:
+          merchantCouponWithoutIdentifier.percentageDiscount.toNumber(),
+        ...(upgradeFromPurchase && {
+          fixedDiscount: fixedDiscountForUpgrade,
+        }),
+      }),
+      appliedMerchantCoupon: merchantCouponWithoutIdentifier,
+      ...(upgradeFromPurchase && {
+        upgradeFromPurchaseId,
+        upgradeFromPurchase,
+        upgradedProduct,
+      }),
+    }
```

Comment:
if they have any restricted/ppp purchase and are upgrading, just apply PPP instead of giving them a branching choice here. It's a safe assumption.

This probably leaves a hole where they want to upgrade to the unrestricted version to remove regional restriction, which DOES happen, but is rarer.